File: src/BattleSim/Status.lua
-- Gen1 status: the before-move gates and the end-of-turn residuals.
--
-- Two families of function live here, and they fire at opposite ends of a
-- turn:
--
--   * the *gates* (sleep, freeze, paralysis, confusion) run before the chosen
--     move and answer one question -- does this side act at all;
--   * the *residuals* (burn, poison, toxic) run after, and answer how much HP
--     the status costs.
--
-- The quirks kept deliberately, each because a mediated fight has to resolve
-- the way the players' own copies would:
--
--   * **Sleep loses the waking turn.**  The counter is decremented *before*
--     the move, and the turn it reaches zero is still spent waking up.  So a
--     3-turn sleep costs three turns, not two.
--   * **Freeze never self-thaws.**  There is no thaw roll in Gen1 at all; only
--     a fire-type hit or an item ends it, so `thawed` is false for every roll
--     and the field exists only to say so out loud.
--   * **Confusion decrements first and snaps out at zero**, and the turn it
--     snaps out on is a free one -- unlike sleep.  Otherwise it self-hits on
--     roll < 128, and the self-hit is a typeless 40-power physical hit using
--     the confused side's *own* attack against its *own* defence, with no
--     STAB, no type chart, no crit, and a fixed max roll.
--   * **Paralysis** full-stops on roll < 63 (63/256, not the 25% the manual
--     claims) and quarters speed.
--
-- Every "a sixteenth" is max(1, floor(maxHp/16)), so a 10 HP monster still
-- takes 1 rather than 0 -- burn and poison can always eventually kill.
--
-- Nothing here raises; no love, no engine modules, no mod facade.  The one
-- dependency is the sibling damage pipeline, which the confusion self-hit
-- routes through rather than reimplementing: two damage formulas in one sim
-- would drift on the first fix to either.

local need = ...
local Damage = need("BattleSim/Damage")

local M = {}

local floor, max = math.floor, math.max

-- 63/256 before the move; 128/256 for a confusion self-hit.
M.PARALYSIS_STOP_ROLL = 63
M.CONFUSION_HIT_ROLL = 128
M.CONFUSION_POWER = 40
M.RESIDUAL_DIVISOR = 16

local function int(value, fallback)
  local n = tonumber(value)
  if not n or n ~= n then return fallback end
  return floor(n)
end

-- ------- before-move gates

-- A sleeping side never acts, so `canMove` is false on every branch; the
-- interesting return is `wokeUp`, which is what the narration hangs off.
--
-- A counter already at zero means "not asleep": the gate passes rather than
-- decrementing into negatives, so a caller that asks the wrong question gets a
-- harmless answer instead of a monster stuck asleep forever.
function M.sleepTick(turnsRemaining)
  local left = max(0, int(turnsRemaining, 0))
  if left == 0 then
    return { turnsRemaining = 0, wokeUp = false, canMove = true }
  end
  left = left - 1
  return { turnsRemaining = left, wokeUp = left == 0, canMove = false }
end

function M.sleepCanMove(turnsRemaining)
  return M.sleepTick(turnsRemaining).canMove
end

-- The roll is accepted and ignored: there is no self-thaw in Gen1, and taking
-- the argument keeps the gate's shape uniform for the turn machine that will
-- call all four of these through one dispatcher.
function M.freezeTick(_roll)
  return { thawed = false, canMove = false }
end

function M.freezeCanMove(_roll)
  return false
end

function M.paralysisStop(roll)
  return int(roll, 255) < M.PARALYSIS_STOP_ROLL
end

function M.paralysisTick(roll)
  local stopped = M.paralysisStop(roll)
  return { fullyParalyzed = stopped, canMove = not stopped }
end

-- max(1, floor(speed/4)) -- the floor matters at low speed, where a straight
-- quarter would reach zero and the turn order would stop being defined.
function M.paralysisSpeed(speed)
  return max(1, floor(max(0, int(speed, 1)) / 4))
end

-- state: { turnsRemaining, level, attack | atk, defense | def }
--
-- The self-hit damage is computed here rather than left to the caller because
-- the parameters that make it typeless are part of the *rule*, not the
-- caller's choice: get one of them wrong and confusion hurts differently on
-- the two runtimes.
function M.confusionTick(state, roll)
  state = state or {}
  local left = max(0, int(state.turnsRemaining, 0))
  if left > 0 then left = left - 1 end

  if left == 0 then
    return { turnsRemaining = 0, snappedOut = true, selfHit = false,
             canMove = true, selfDamage = 0 }
  end

  if not M.confusionSelfHit(roll) then
    return { turnsRemaining = left, snappedOut = false, selfHit = false,
             canMove = true, selfDamage = 0 }
  end

  local hit = Damage.compute(
    { level = state.level, attack = state.attack, atk = state.atk },
    { defense = state.defense, def = state.def },
    { power = M.CONFUSION_POWER },
    { crit = false, stab = false, typeEffect = { 100 }, roll = Damage.ROLL_MAX })

  return { turnsRemaining = left, snappedOut = false, selfHit = true,
           canMove = false, selfDamage = hit.damage or 0 }
end

function M.confusionSelfHit(roll)
  return int(roll, 255) < M.CONFUSION_HIT_ROLL
end

-- ------- residuals

local function sixteenth(maxHp)
  return max(1, floor(max(0, int(maxHp, 0)) / M.RESIDUAL_DIVISOR))
end

function M.residualBurn(maxHp)
  return sixteenth(maxHp)
end

function M.residualPoison(maxHp)
  return sixteenth(maxHp)
end

-- Toxic stacks by multiplying the same sixteenth by the badly-poisoned
-- counter, which the turn machine increments once per tick.
function M.residualToxic(maxHp, toxicCounter)
  return sixteenth(maxHp) * max(1, int(toxicCounter, 1))
end

function M.burnAttack(attack)
  return max(1, floor(max(0, int(attack, 1)) / 2))
end

-- ------- dispatcher
--
-- One entry point for the turn machine, so the call order lives in one place
-- rather than being reassembled at each of the sites that gate a move.
-- Returns the same tables the individual gates do, plus the status name, and
-- nil for a status with no before-move gate at all (burn, poison, toxic) so
-- the caller can tell "passed the gate" from "there was no gate".
function M.beforeMove(state, roll)
  state = state or {}
  local status = state.status
  if status == "sleep" then
    local out = M.sleepTick(state.turnsRemaining); out.status = status; return out
  elseif status == "freeze" then
    local out = M.freezeTick(roll); out.status = status; return out
  elseif status == "paralysis" then
    local out = M.paralysisTick(roll); out.status = status; return out
  elseif status == "confusion" then
    local out = M.confusionTick(state, roll); out.status = status; return out
  end
  return nil
end

-- The end-of-turn half of the same seam.  nil when the status costs no HP.
function M.residual(state)
  state = state or {}
  local status = state.status
  if status == "burn" then return M.residualBurn(state.maxHp) end
  if status == "poison" then return M.residualPoison(state.maxHp) end
  if status == "toxic" then
    return M.residualToxic(state.maxHp, state.toxicCounter)
  end
  return nil
end

return M

File: server/lib/battle/Status.js
'use strict';

/*
 * Gen1 status: what stops a turn, what it costs at the end of one, and what it
 * does to a stat while it lasts.
 *
 * Node half of src/BattleSim/Status.lua; same parity contract as Damage.js,
 * and the function names are the Lua twin's on purpose, so the two files can
 * be read side by side by whoever is chasing a disagreement.
 *
 * Status is where a mediated battle is easiest to get subtly wrong, because
 * most of it is off-turn bookkeeping the player never sees a number for. The
 * quirks kept here on purpose:
 *
 *   * Sleep decrements *before* the move and the waking turn is still lost, so
 *     a 1-turn sleep costs a turn rather than none.
 *   * Freeze never self-thaws. In Gen1 only a fire move or a healing item ends
 *     it, so `thawed` is false for every roll and the roll is accepted only so
 *     callers do not have to special-case which statuses want one.
 *   * Paralysis stops the turn 63/256 of the time -- not 25% -- and quarters
 *     speed with a floor of 1.
 *   * Confusion decrements first and snaps out at zero, so the turn a mon
 *     comes out of confusion is a turn it gets to act.
 *
 * The confusion self-hit is a typeless 40-power move using the confused side's
 * own attack against its own defence, with no STAB, no type effectiveness, no
 * crit and no random roll (the roll is pinned at maximum). It runs through the
 * same Damage.compute as everything else rather than getting its own
 * arithmetic, so it cannot drift away from the main formula.
 *
 * Edge inputs match Lua: paralysisStop(undefined) coerces the roll to 255
 * (false), and confusionTick forwards atk/def aliases into Damage.compute.
 */

const Damage = require('./Damage.js');

const idiv = (a, b) => Math.floor(a / b);

const PARALYSIS_STOP_ROLL = 63; // out of 256
const CONFUSION_HIT_ROLL = 128; // out of 256
const CONFUSION_POWER = 40;
const RESIDUAL_DIVISOR = 16;

function int(value, fallback) {
  const n = Number(value);
  if (value === null || value === undefined || Number.isNaN(n)) return fallback;
  if (typeof value === 'boolean') return fallback;
  return Math.floor(n);
}

// ------- before-move gates

// A counter already at zero means "not asleep": the gate passes rather than
// decrementing into negatives, so a caller that asks the wrong question gets a
// harmless answer instead of a monster stuck asleep forever.
function sleepTick(turnsRemaining) {
  const left = Math.max(0, int(turnsRemaining, 0));
  if (left === 0) return { turnsRemaining: 0, wokeUp: false, canMove: true };
  return { turnsRemaining: left - 1, wokeUp: left - 1 === 0, canMove: false };
}

function sleepCanMove(turnsRemaining) {
  return sleepTick(turnsRemaining).canMove;
}

// The roll is accepted and ignored: there is no self-thaw in Gen1, and taking
// the argument keeps the gate's shape uniform for the turn machine that will
// call all four of these through one dispatcher.
function freezeTick(_roll) {
  return { thawed: false, canMove: false };
}

function freezeCanMove(_roll) {
  return false;
}

// Missing roll → 255, matching Lua int(roll, 255): 255 < 63 is false.
function paralysisStop(roll) {
  return int(roll, 255) < PARALYSIS_STOP_ROLL;
}

function paralysisTick(roll) {
  const stopped = paralysisStop(roll);
  return { fullyParalyzed: stopped, canMove: !stopped };
}

// max(1, floor(speed/4)) -- the floor matters at low speed, where a straight
// quarter would reach zero and the turn order would stop being defined.
function paralysisSpeed(speed) {
  return Math.max(1, idiv(Math.max(0, int(speed, 1)), 4));
}

function confusionSelfHit(roll) {
  return int(roll, 255) < CONFUSION_HIT_ROLL;
}

// state: { turnsRemaining, level, attack | atk, defense | def }
//
// The self-hit damage is computed here rather than left to the caller because
// the parameters that make it typeless are part of the *rule*, not the
// caller's choice: get one of them wrong and confusion hurts differently on
// the two runtimes.
function confusionTick(state, roll) {
  const s = state || {};
  let left = Math.max(0, int(s.turnsRemaining, 0));
  if (left > 0) left -= 1;

  if (left === 0) {
    return { turnsRemaining: 0, snappedOut: true, selfHit: false, canMove: true, selfDamage: 0 };
  }
  if (!confusionSelfHit(roll)) {
    return { turnsRemaining: left, snappedOut: false, selfHit: false, canMove: true, selfDamage: 0 };
  }

  // Forward atk/def aliases the way Lua does, so Damage.compute's stat() reads
  // whichever spelling the caller had.
  const hit = Damage.compute({
    level: s.level,
    power: CONFUSION_POWER,
    attack: s.attack,
    atk: s.atk,
    defense: s.defense,
    def: s.def,
    crit: false,
    stab: false,
    typeEffect: [100],
    roll: Damage.ROLL_MAX,
  });

  return {
    turnsRemaining: left,
    snappedOut: false,
    selfHit: true,
    canMove: false,
    selfDamage: hit.damage || 0,
  };
}

// ------- residuals

function sixteenth(maxHp) {
  return Math.max(1, idiv(Math.max(0, int(maxHp, 0)), RESIDUAL_DIVISOR));
}

function residualBurn(maxHp) {
  return sixteenth(maxHp);
}

function residualPoison(maxHp) {
  return sixteenth(maxHp);
}

// Toxic stacks by multiplying the same sixteenth by the badly-poisoned
// counter, which the turn machine increments once per tick.
function residualToxic(maxHp, toxicCounter) {
  return sixteenth(maxHp) * Math.max(1, int(toxicCounter, 1));
}

function burnAttack(attack) {
  return Math.max(1, idiv(Math.max(0, int(attack, 1)), 2));
}

// ------- dispatchers

/*
 * One entry point for the turn machine, so the call order lives in one place
 * rather than being reassembled at each of the sites that gate a move.
 * Returns the same objects the individual gates do, plus the status name, and
 * null for a status with no before-move gate at all (burn, poison, toxic) so
 * the caller can tell "passed the gate" from "there was no gate".
 */
function beforeMove(state, roll) {
  const s = state || {};
  let out = null;
  switch (s.status) {
    case 'sleep': out = sleepTick(s.turnsRemaining); break;
    case 'freeze': out = freezeTick(roll); break;
    case 'paralysis': out = paralysisTick(roll); break;
    case 'confusion': out = confusionTick(s, roll); break;
    default: return null;
  }
  out.status = s.status;
  return out;
}

// The end-of-turn half of the same seam. null when the status costs no HP.
function residual(state) {
  const s = state || {};
  if (s.status === 'burn') return residualBurn(s.maxHp);
  if (s.status === 'poison') return residualPoison(s.maxHp);
  if (s.status === 'toxic') return residualToxic(s.maxHp, s.toxicCounter);
  return null;
}

/*
 * evaluate(record) -- one door for a fixture-shaped record.
 *
 * The turn machine calls `beforeMove` and `residual`; this exists for the
 * parity suites, which drive a flat table where the fields present decide
 * which of the seams above a case is asking about (a paralysis record with a
 * `speed` is asking about the stat, one with a `roll` about the gate). Returns
 * null for a combination this module has no answer for, rather than guessing.
 */
function evaluate(record) {
  const r = record || {};
  switch (r.status) {
    case 'paralysis':
      if (r.speed !== undefined) return { speed: paralysisSpeed(r.speed) };
      return beforeMove(r, r.roll);
    case 'burn':
      if (r.attack !== undefined) return { attack: burnAttack(r.attack) };
      return { residualDamage: residualBurn(r.maxHp) };
    case 'poison':
    case 'toxic':
      return { residualDamage: residual(r) };
    default:
      return beforeMove(r, r.roll);
  }
}

module.exports = {
  beforeMove,
  residual,
  evaluate,
  sleepTick,
  sleepCanMove,
  freezeTick,
  freezeCanMove,
  paralysisStop,
  paralysisTick,
  paralysisSpeed,
  confusionSelfHit,
  confusionTick,
  residualBurn,
  residualPoison,
  residualToxic,
  burnAttack,
  int,
  PARALYSIS_STOP_ROLL,
  CONFUSION_HIT_ROLL,
  CONFUSION_POWER,
  RESIDUAL_DIVISOR,
};
